### PR TITLE
[SPARK-26493][SQL] Allow multiple spark.sql.extensions

### PR DESCRIPTION
**Pull request by jamisonbennett**
## What changes were proposed in this pull request?

Allow multiple spark.sql.extensions to be specified in the
configuration.

## How was this patch tested?

New tests are added.

### CODE REVIEW

**Comment by jamisonbennett** [comment]
This is my original work and I license the work to the project under the project’s open source license.
 I have signed an ICLA with apache.org.

**Comment by RussellSpitzer** [comment]
Actually how does this work from a spark defaults perspective, does a comma separated string work?

**Comment by gatorsmile** [comment]
@RussellSpitzer See this example. 

```Scala
    val conf = new SparkConf()
    val seq = ConfigBuilder(testKey("seq")).stringConf.toSequence.createWithDefault(Seq())
    conf.set(seq.key, "1,,2, 3 , , 4")
    assert(conf.get(seq) === Seq("1", "2", "3", "4"))
    conf.set(seq, Seq("1", "2"))
    assert(conf.get(seq) === Seq("1", "2"))
```

**Comment by dongjoon-hyun** [comment]
This PR looks useful. Thank you for your first contribution, @jamisonbennett .

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 1]
```diff
@@ -114,6 +124,25 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
...
+        classOf[MyExtensions].getCanonicalName,
+        classOf[MyExtensions2].getCanonicalName).mkString(","))
+      .getOrCreate()
+    try {
+      assert(session.sessionState.planner.strategies.contains(MySparkStrategy(session)))
+      assert(session.sessionState.analyzer.extendedResolutionRules.contains(MyRule(session)))
+      assert(session.sessionState.functionRegistry
+        .lookupFunction(MyExtensions.myFunction._1).isDefined)
+      assert(session.sessionState.functionRegistry
+        .lookupFunction(MyExtensions2.myFunction._1).isDefined)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: So, now we have multiple extension registrations. The order of extension names might have side-effects.
- Can we have a test case for duplicated extension names? `MyExtension2` and `MyExtension2`?
- Can we have a negative test case for function name conflicts? `MyExtension2.myFunction` and `MyExtension3.myFunction`?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: I think the order matters, but we need to discuss and document the behavior when we have name conflicts. 

For example, the same rule will be added twice in `extendedResolutionRules`. Is it desired?

```Scala
class MyExtensions extends (SparkSessionExtensions => Unit) {
  def apply(e: SparkSessionExtensions): Unit = {
    e.injectResolutionRule(MyRule)
  }
}

class MyExtensions2 extends (SparkSessionExtensions => Unit) {
  def apply(e: SparkSessionExtensions): Unit = {
    e.injectResolutionRule(MyRule)
  }
}

```

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Yep. If there is no reason to allow that, we had better disallow that by design before this PR.

Comment (hvanhovell) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: There are use cases where you want to execute rules in a certain order. So I think it is reasonable to add the same rule multiple times. If you want more control you could even create 'micro' optimizer batches by calling multiple rules from one rule.

I think this is more a matter of proper documentation than one where we should explicitly block things. Also note that this is a pretty advanced feature and by this stage users are expected to know what they are doing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 2]
```diff
@@ -99,9 +99,10 @@ object StaticSQLConf {
       .createWithDefault(false)
 
   val SPARK_SESSION_EXTENSIONS = buildStaticConf("spark.sql.extensions")
-    .doc("Name of the class used to configure Spark Session extensions. The class should " +
+    .doc("List of the class names used to configure Spark Session extensions. The classes should " +
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: Please document in what order rules from multiple extensions are executed, how listeners are invoked and how functions are registered (last one wins?). Be sure to cover what happens if you add duplicate listeners/rules/functions etc...

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 1 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Prior to this change, it was possible to programmatically register multiple extensions but it was not possible to do so through the spark.sql.extensions configuration. Although it wasn't documented/tested until this pull request. E.g. The following works without this pull request:
```
SparkSession.builder()
  .master("..")
  .withExtensions(sparkSessionExtensions1)
  .withExtensions(sparkSessionExtensions2)
  .getOrCreate()
```

So I think conflicting function names are already currently possible (but not documented). In the following cases:
1. Conflicting function names are registered by calling `.withExtenions()` multiple times
2. An extension accidentally registers a function that was already registered with the builtin functions
3. An extension accidentally registers a function multiple times by calling `injectFunction(myFunction)`

As for the order, it looks to me like the last function to be stored with conflicting names is the one which is retrieved:
```
class SimpleFunctionRegistry extends FunctionRegistry {

  @GuardedBy("this")
  private val functionBuilders =
    new mutable.HashMap[FunctionIdentifier, (ExpressionInfo, FunctionBuilder)]
  override def registerFunction(
      name: FunctionIdentifier,
      info: ExpressionInfo,
      builder: FunctionBuilder): Unit = synchronized {
    functionBuilders.put(normalizeFuncName(name), (info, builder))
  }
```

I will update this PR to document what happens in order of operations and conflicts. If we need to explicitly block duplicates functions from being registered, I can temporarily drop this PR and see about making those changes first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: A suggestion of update the comment is replace 'List of the class names used to configure Spark Session extensions.  The classes should implement Function1[SparkSessionExtension, Unit], and must have a no-args constructor.' to 'A comma-separated list of classes that implement Function1[SparkSessionExtension, Unit] used to configure Spark Session extensions. The classes must have a no-args constructor.'

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 1 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Can we have a test case for duplicated extension names? Done
Can we have a negative test case for function name conflicts? MyExtension2.myFunction and MyExtension3.myFunction? Done

I added documentation for the behavior.
I added a warning message if a registered function is replaced.
I added a test case for the ordering.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 2 above)
Comment (jamisonbennett) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: I updated the documentation as suggested with respect to the ordering. I think the comment about "listeners" is related to code that is close in proximity to the code for this pull request, but it is a different configuration item. So I didn't update the `spark.sql.queryExecutionListeners` documentation. If you think that documentation should be updated, let me know if I should include it as a part of this pull request or as a separate pull request.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 3]
```diff
@@ -103,7 +104,10 @@ class SimpleFunctionRegistry extends FunctionRegistry {
       name: FunctionIdentifier,
       info: ExpressionInfo,
       builder: FunctionBuilder): Unit = synchronized {
-    functionBuilders.put(normalizeFuncName(name), (info, builder))
+    val normalizedName = normalizeFuncName(name)
+    if (functionBuilders.put(normalizedName, (info, builder)).isDefined) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: It would be great if we can check if the new function and the old function are different. This will help to increase the signal of the error message.

Comment (jamisonbennett) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I added a check which which will only log if different function objects are registered. The "allow an extension to be duplicated" unit tests that I previously added registers the same object twice. This test no longer prints the warning. The "use the last registered function name when there are duplicates" unit tests that I previously added registers different functions with the same name. This test prints the warning.

**Comment by bersprockets** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]
Looks okay. It's unstable + experimental FWIW.

**Comment by felixcheung** [comment]
@hvanhovell @gatorsmile ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 4]
```diff
@@ -99,9 +99,15 @@ object StaticSQLConf {
...
 
   val SPARK_SESSION_EXTENSIONS = buildStaticConf("spark.sql.extensions")
-    .doc("Name of the class used to configure Spark Session extensions. The class should " +
-      "implement Function1[SparkSessionExtension, Unit], and must have a no-args constructor.")
+    .doc("A comma-separated list of classes that implement " +
+      "Function1[SparkSessionExtension, Unit] used to configure Spark Session extensions. The " +
+      "classes must have a no-args constructor. If multiple extensions are specified, they are " +
+      "applied in the specified order. For the case of rules and planner strategies, they are " +
+      "applied in the specified order. For the case of parsers, the last parser is used and each " +
+      "parser can delegate to its predecessor. For the case of function name conflicts, the last " +
```
[the remark this answers is not in the document]
Comment (jamisonbennett) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: I added an example to the documentation of `SparkSessionExtensions`.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 5]
```diff
@@ -78,6 +86,14 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
     }
   }
 
+  test("inject multiple rules") {
+    withSession(Seq(_.injectOptimizerRule(MyRule),
+      _.injectPlannerStrategy(MySparkStrategy))) { session =>
```
Comment (vanzin) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit: indent more, otherwise it blends with the code block below

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 6]
```diff
@@ -108,12 +124,86 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
...
+        Seq(MySparkStrategy2(session), MySparkStrategy(session))))
+      val orderedRules = Seq(MyRule2(session), MyRule(session))
+      val orderedCheckRules = Seq(MyCheckRule2(session), MyCheckRule(session))
+      val parser = MyParser(session, CatalystSqlParser)
+      assert(session.sessionState.analyzer.extendedResolutionRules.containsSlice(orderedRules))
+      assert(session.sessionState.analyzer.postHocResolutionRules.containsSlice(orderedRules))
+      assert(session.sessionState.analyzer.extendedCheckRules.containsSlice(orderedCheckRules))
+      assert(session.sessionState.optimizer.batches.flatMap(_.rules).filter(orderedRules.contains)
+        .containsSlice(orderedRules ++ orderedRules)) // The optimizer rules are duplicated
+      assert(session.sessionState.sqlParser == parser)
```
Comment (vanzin) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: In all these asserts, use `===` and `!==`.

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: That's actually arguable, Vanzin. Some people prefer `===` whereas some prefer `==`. `===` doesn't look always reporting a better error message give my tests. See also https://github.com/databricks/scala-style-guide/issues/36.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 7]
```diff
@@ -103,7 +104,13 @@ class SimpleFunctionRegistry extends FunctionRegistry {
       name: FunctionIdentifier,
       info: ExpressionInfo,
       builder: FunctionBuilder): Unit = synchronized {
-    functionBuilders.put(normalizeFuncName(name), (info, builder))
+    val normalizedName = normalizeFuncName(name)
+    val newFunction = (info, builder)
+    functionBuilders.put(normalizedName, newFunction) match {
+      case Some(previousFunction) if previousFunction != newFunction =>
+        logWarning(s"The function $normalizedName replaced a previously registered function.")
+      case _ => Unit
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: This is returning Unit type object. This can be just removed.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 8]
```diff
@@ -151,14 +241,63 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
 object MyExtensions {
 
   val myFunction = (FunctionIdentifier("myFunction"),
-    new ExpressionInfo("noClass", "myDb", "myFunction", "usage", "extended usage" ),
+    new ExpressionInfo("noClass", "myDb", "myFunction", "usage", "extended usage"),
     (myArgs: Seq[Expression]) => Literal(5, IntegerType))
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: I would remove this newline. Looks unrelated.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 9]
```diff
@@ -151,14 +241,63 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
...
+}
+
+class MyExtensions2 extends (SparkSessionExtensions => Unit) {
+  def apply(e: SparkSessionExtensions): Unit = {
+    e.injectPlannerStrategy(MySparkStrategy2)
+    e.injectResolutionRule(MyRule2)
+    e.injectPostHocResolutionRule(MyRule2)
+    e.injectCheckRule(MyCheckRule2)
+    e.injectOptimizerRule(MyRule2)
+    e.injectParser((_, _) => CatalystSqlParser)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit: `e.injectParser((_: SparkSession, _: ParserInterface) => CatalystSqlParser)`

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 10]
```diff
@@ -151,14 +241,63 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
 object MyExtensions {
 
   val myFunction = (FunctionIdentifier("myFunction"),
-    new ExpressionInfo("noClass", "myDb", "myFunction", "usage", "extended usage" ),
+    new ExpressionInfo("noClass", "myDb", "myFunction", "usage", "extended usage"),
     (myArgs: Seq[Expression]) => Literal(5, IntegerType))
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit: `(_: Seq[Expression]) => Literal(5, IntegerType))`

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 11]
```diff
@@ -151,14 +241,63 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
...
+    e.injectOptimizerRule(MyRule2)
+    e.injectParser((_, _) => CatalystSqlParser)
+    e.injectFunction(MyExtensions2.myFunction)
+  }
+}
+
+object MyExtensions2Duplicate {
+
+  val myFunction = (FunctionIdentifier("myFunction2"),
+    new ExpressionInfo("noClass", "myDb", "myFunction2", "usage", "last wins" ),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit `"last wins"` -> `"extended usage"`

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit: `" )` -> `")`

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 12]
```diff
@@ -151,14 +241,63 @@ case class MyParser(spark: SparkSession, delegate: ParserInterface) extends Pars
...
+}
+
+case class MySparkStrategy2(spark: SparkSession) extends SparkStrategy {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = Seq.empty
+}
+
+object MyExtensions2 {
+
+  val myFunction = (FunctionIdentifier("myFunction2"),
+    new ExpressionInfo("noClass", "myDb", "myFunction2", "usage", "extended usage" ),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: nit: `" )` -> `")`

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala [hunk 13]
```diff
@@ -61,6 +61,26 @@ import org.apache.spark.sql.catalyst.rules.Rule
  *     .getOrCreate()
  * }}}
  *
+ * The extensions can also be used by setting the Spark SQL configuration property
+ * spark.sql.extensions, for example:
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala]: nit: spark.sql.extensions -> `'spark.sql.extensions'`

Shall we also mention multiple rule can be set via comma separated?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 7 above)
Comment (jamisonbennett) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I made a change here which I think was what you were recommending. I didn't want to remove the `case _ =>` otherwise I think it may result in a non-exhaustive match exception. If you wanted me to remove the entire match statement, just let me know.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 8 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 9 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: I also made the suggested change in 2 other places in this file so that it is consistent.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 10 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 11 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: I made both changes and also updated one of the tests to validate the `ExpressionInfo` object rather than the extended usage text.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 12 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala (hunk 13 above)
Comment (jamisonbennett) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala]: I added the recommended documentation updates and also added code marks around `withExtensions` and fixed the typo there. I didn't think it was necessary to provide the example of using the comma-separated string but I did note it in the documentation.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 6 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Based on [databricks/scala-style-guide#36](https://github.com/databricks/scala-style-guide/issues/36), it looks like `==` might now be preferred over `===`. For what its worth, it seems that in the cases for this test `==` produces reasonable error messages such as `MyParser(org.apache.spark.sql.SparkSession@6e8a9c30,org.apache.spark.sql.catalyst.parser.CatalystSqlParser$@5d01ea21) did not equal IntentionalErrorThatIInsertedHere` and `2 did not equal 3`. So please let me know if there is newer guidance to use `===` and I can make the changes.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala (hunk 6 above)
Comment (jamisonbennett) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: Thanks for the follow up, I was not aware of that difference. I updated the tests to use `===` and `!==` as originally recommended.

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala]: @vanzin, where does it say we use `===`? If it's practically yes, let's document it.

**Comment by HyukjinKwon** [comment]
Merged to master.

For `===` vs `==`, if that's not documented, let's document in databricks/scala-style-guide which is an official style guide to refer. If that's not documented, I think both are okay. Both are already being used here and there.
